File: Indexed Search/Hash Based Indexed Search/Extendible Hashing/ExtendibleHashing.js
/*
====================================
Bucket Structure
------------------------------------
localDepth:
Number of directory bits
used by this bucket
====================================
*/

class Bucket {

    constructor(
        localDepth,
        capacity
    ) {

        this.localDepth =
            localDepth;

        this.capacity =
            capacity;

        this.keys = [];
    }

    isFull() {

        return (
            this.keys.length
            >=
            this.capacity
        );
    }

    contains(
        key
    ) {

        return (
            this.keys.includes(
                key
            )
        );
    }

    insert(
        key
    ) {

        this.keys.push(
            key
        );
    }

    delete(
        key
    ) {

        const index =
            this.keys.indexOf(
                key
            );

        if (
            index
            !==
            -1
        ) {

            this.keys.splice(
                index,
                1
            );
        }
    }
}



/*
====================================
Dynamic Hashing
(Extendible Hashing)
====================================
*/

class DynamicHashing {

    static BUCKET_SIZE = 3;

    constructor() {

        this.globalDepth =
            1;

        this.directory =
            [
                new Bucket(
                    1,
                    DynamicHashing.BUCKET_SIZE
                ),

                new Bucket(
                    1,
                    DynamicHashing.BUCKET_SIZE
                )
            ];
    }



    /*
    Hash Formula

    index =
    key &
    ((1<<globalDepth)-1)
    */

    getIndex(
        key
    ) {

        return (
            key
            &
            (
                (
                    1
                    <<
                    this.globalDepth
                )
                -
                1
            )
        );
    }



    /*
    Expand directory

    New Size:

    2^globalDepth
    */

    expandDirectory() {

        const oldSize =
            this.directory
            .length;

        for (
            let i = 0;
            i < oldSize;
            i++
        ) {

            this.directory
                .push(
                    this.directory[i]
                );
        }

        this.globalDepth++;
    }



    /*
    Split overflowing bucket
    */

    splitBucket(
        index
    ) {

        const oldBucket =
            this.directory[
                index
            ];

        if (
            oldBucket.localDepth
            ===
            this.globalDepth
        ) {

            this.expandDirectory();
        }


        const newBucket =
            new Bucket(
                oldBucket.localDepth + 1,
                DynamicHashing.BUCKET_SIZE
            );


        oldBucket.localDepth++;


        const splitBit =
            (
                1
                <<
                (
                    oldBucket.localDepth
                    -
                    1
                )
            );


        for (
            let i = 0;
            i < this.directory.length;
            i++
        ) {

            if (
                this.directory[i]
                ===
                oldBucket
                &&
                (
                    i
                    &
                    splitBit
                )
            ) {

                this.directory[i] =
                    newBucket;
            }
        }


        const oldValues =
            [
                ...oldBucket.keys
            ];


        oldBucket.keys = [];


        for (
            const value
            of
            oldValues
        ) {

            const newIndex =
                this.getIndex(
                    value
                );

            this.directory[
                newIndex
            ]
            .insert(
                value
            );
        }
    }



    /*
    Insert key
    */

    insert(
        key
    ) {

        while (
            true
        ) {

            const index =
                this.getIndex(
                    key
                );

            const bucket =
                this.directory[
                    index
                ];

            if (
                bucket.contains(
                    key
                )
            ) {

                return;
            }

            if (
                !bucket.isFull()
            ) {

                bucket.insert(
                    key
                );

                return;
            }

            this.splitBucket(
                index
            );
        }
    }



    /*
    Search
    */

    search(
        key
    ) {

        const index =
            this.getIndex(
                key
            );

        return (
            this.directory[
                index
            ]
            .contains(
                key
            )
        );
    }



    /*
    Delete
    */

    delete(
        key
    ) {

        const index =
            this.getIndex(
                key
            );

        this.directory[
            index
        ]
        .delete(
            key
        );
    }



    /*
    Display
    */

    display() {

        console.log(
            "\nGlobal Depth:",
            this.globalDepth
        );

        console.log();

        for (
            let i = 0;
            i < this.directory.length;
            i++
        ) {

            console.log(
                `Dir[${i}] ->`,
                [
                    ...this.directory[i].keys
                ]
            );
        }
    }
}



/*
====================================
Driver
====================================
*/

const table =
    new DynamicHashing();


[
    10,
    22,
    14,
    30,
    42,
    54,
    66
]
.forEach(
    value =>
    table.insert(
        value
    )
);


table.display();


console.log(
    "\nSearch 42:",
    table.search(
        42
    )
);


table.delete(
    42
);


console.log(
    "\nAfter Delete:"
);


table.display();

File: Indexed Search/Hash Based Indexed Search/Extendible Hashing/ExtendibleHashing.py
class Bucket:
    """
    Bucket used in Extendible Hashing

    local_depth:
    Number of directory bits
    this bucket recognizes
    """

    def __init__(self, local_depth, capacity):

        self.local_depth = local_depth
        self.capacity = capacity
        self.keys = []

    def is_full(self):
        return len(self.keys) >= self.capacity

    def contains(self, key):
        return key in self.keys

    def insert(self, key):
        self.keys.append(key)

    def delete(self, key):

        if key in self.keys:
            self.keys.remove(key)


class DynamicHashing:
    """
    Dynamic Hashing
    (Extendible Hashing)
    """

    BUCKET_SIZE = 3

    def __init__(self):

        self.global_depth = 1

        self.directory = [
            Bucket(1, self.BUCKET_SIZE),
            Bucket(1, self.BUCKET_SIZE)
        ]

    """
    Hash Formula

    index =
    key &
    ((1 << global_depth)-1)
    """

    def get_index(self, key):

        return (
            key
            &
            (
                (1 << self.global_depth)
                - 1
            )
        )

    """
    Double directory
    """

    def expand_directory(self):

        old_size = len(self.directory)

        for i in range(old_size):

            self.directory.append(
                self.directory[i]
            )

        self.global_depth += 1

    """
    Split bucket
    """

    def split_bucket(self, index):

        old_bucket = self.directory[index]

        if (
            old_bucket.local_depth
            ==
            self.global_depth
        ):

            self.expand_directory()

        new_bucket = Bucket(
            old_bucket.local_depth + 1,
            self.BUCKET_SIZE
        )

        old_bucket.local_depth += 1

        split_bit = (
            1
            <<
            (
                old_bucket.local_depth
                - 1
            )
        )

        for i in range(
            len(self.directory)
        ):

            if (
                self.directory[i]
                is old_bucket
                and
                (
                    i & split_bit
                )
            ):

                self.directory[i] = new_bucket

        old_values = list(
            old_bucket.keys
        )

        old_bucket.keys.clear()

        for value in old_values:

            new_index = (
                self.get_index(
                    value
                )
            )

            self.directory[
                new_index
            ].insert(
                value
            )

    """
    Insert key
    """

    def insert(self, key):

        while True:

            index = (
                self.get_index(
                    key
                )
            )

            bucket = (
                self.directory[
                    index
                ]
            )

            if (
                bucket.contains(
                    key
                )
            ):

                return

            if (
                not
                bucket.is_full()
            ):

                bucket.insert(
                    key
                )

                return

            self.split_bucket(
                index
            )

    """
    Search key
    """

    def search(self, key):

        index = (
            self.get_index(
                key
            )
        )

        return (
            self.directory[
                index
            ]
            .contains(
                key
            )
        )

    """
    Delete key
    """

    def delete(self, key):

        index = (
            self.get_index(
                key
            )
        )

        self.directory[
            index
        ].delete(
            key
        )

    """
    Display table
    """

    def display(self):

        print(
            "\nGlobal Depth:",
            self.global_depth
        )

        print()

        for i, bucket in enumerate(
            self.directory
        ):

            print(
                f"Dir[{i}] ->",
                bucket.keys
            )


# ------------------------
# Driver
# ------------------------

if __name__ == "__main__":

    table = DynamicHashing()

    values = [
        10,
        22,
        14,
        30,
        42,
        54,
        66
    ]

    for value in values:
        table.insert(
            value
        )

    table.display()

    print(
        "\nSearch 42:",
        table.search(
            42
        )
    )

    table.delete(
        42
    )

    print(
        "\nAfter Delete:"
    )

    table.display()
